tools/autograd/templates/variable_factories.h

```diff
@@ -22,61 +22,96 @@ using at::DimnameList;
+        scalar_type_(at::k##S),                                                          \
+        type_(ListInitTensorType::Vector) {                                              \
+      vec_.reserve(values.size());                                                       \
+      for (const auto& elem : values) {                                                  \
+        vec_.push_back(ListInitTensor(elem));                                            \
+      }                                                                                  \
+    }                                                                                    \
+    ListInitTensor(at::ArrayRef<T> values) :                                             \
+        scalar_(), vec_(),                                                               \
+        /* NOTE: We use `static_cast<int64_t>(...)` here because `at::ArrayRef<T>.size()`
+           returns a value of `size_t` type, and converting it to `int64_t` type is
+           a narrowing conversion. */                                                    \
+        sizes_({static_cast<int64_t>(values.size())}),                                   \
+        scalar_type_(at::k##S),                                                          \
+        type_(ListInitTensorType::Vector) {                                              \
+      vec_.reserve(values.size());                                                       \
+      for (const auto& elem : values) {                                                  \
+        vec_.push_back(ListInitTensor(elem));                                            \
+      }                                                                                  \
+    }
 AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
 #undef TENSOR
-    InitListTensor(std::initializer_list<InitListTensor> init_list) :
+    ListInitTensor(std::initializer_list<ListInitTensor> init_list) :
         scalar_(),
-        init_list_(init_list),
+        vec_(init_list),
         sizes_(),
         scalar_type_(),
-        type_(InitListTensorType::InitList) {
-      TORCH_CHECK(init_list.size() > 0, "Empty init-list is not supported");
-      scalar_type_ = init_list.begin()->scalar_type_;
-      const InitListTensor& first_elem = *(init_list.begin());
-      for (const auto& elem : init_list) {
-        TORCH_CHECK(elem.scalar_type_ == first_elem.scalar_type_,
-          "Expected all elements of the tensor to have the same scalar type: ",
-          first_elem.scalar_type_,
-          ", but got element of scalar type: ",
-          elem.scalar_type_);
-        TORCH_CHECK(elem.sizes_ == first_elem.sizes_,
-          "Expected all sub-lists to have sizes: ",
-          first_elem.sizes_,
-          " (e.g. ", first_elem, "), ",
-          "but got sub-list ",
-          elem,
-          " with sizes: ",
-          elem.sizes_);
+        type_(ListInitTensorType::Vector) {
+      if (init_list.size() > 0) {
```

Comment:
Nit: you can avoid the additional changes (and reduce the number of lines changed) by doing something like:
```
if (init_list.size() == 0) {
  scalar_type_ = c10::ScalarType::Undefined;
  sizes_ = {0};
  return;
}
// existing code here

```